File: src/main/scala/com/mojolly/backchat/zeromq/ZMessage.scala
package com.mojolly.backchat
package zeromq

import java.nio.charset.Charset
import collection.mutable.ListBuffer
import annotation.tailrec

object ZMessage {

  val DELIMITER = Array[Byte]()
  var defaultCharset = Charset.forName("UTF-8")

  def apply(socket: Socket) = {
    @tailrec
    def read(result: ListBuffer[Array[Byte]] = ListBuffer[Array[Byte]]()): List[Array[Byte]] = {
      result += socket.recv(0)
      if (!socket.hasReceiveMore) { result.toList }
      else read(result)
    }
    new ZMessage(read())
  }

  def apply(body: String) = {
    new ZMessage(List(body.getBytes(defaultCharset)))
  }

  def apply(body: String, charset: Charset) = {
    new ZMessage(List(body.getBytes(charset)), charset)
  }

  def apply(bodyParts: String*) = {
    new ZMessage(bodyParts.map(_.getBytes(defaultCharset)).toList)
  }

  def apply(charset: Charset, bodyParts: String*) = {
    new ZMessage(bodyParts.map(_.getBytes(charset)).toList, charset)
  }

  //  def apply(body: Array[Byte]) = {
  //    new ZMessage(List(body))
  //  }
}
class ZMessage(parts: List[Array[Byte]] = Nil, val charset: Charset = ZMessage.defaultCharset) {

  import ZMessage.DELIMITER

  private var _parts = ListBuffer[Array[Byte]](parts: _*)

  def size = _parts.size

  def address = {
    new String(parts.headOption getOrElse DELIMITER, charset)
  }
  def address_=(data: String) { append(data.getBytes(charset)) }
  def body = new String(_parts.lastOption getOrElse DELIMITER, charset)
  def body_=(bd: String) {
    val bdb = bd.getBytes(charset)
    _parts.remove(_parts.size - 1)
    _parts.append(bdb)
  }

  private def get(index: Int) = {
    val ccb = if (_parts.size >= index) {
      _parts(size - index)
    } else DELIMITER
    new String(ccb, charset)
  }
  private def set(index: Int, value: String) {
    val mt = value.getBytes(charset)
    if (size > (index - 1)) {
      _parts.remove(size - index)
      _parts.insert(size - (index - 1), mt)
    } else push(mt)
  }

  def ccid = get(5)

  def ccid_=(mt: String) {
    set(5, mt)
  }

  def target = get(2)

  def target_=(mt: String) {
    set(2, mt)
  }

  def sender = get(3)

  def sender_=(mt: String) {
    set(3, mt)
  }

  def messageType = get(4)

  def messageType_=(mt: String) {
    set(4, mt)
  }

  def append(data: String): ZMessage = {
    append(data.getBytes(charset))
  }

  def push(data: String): ZMessage = {
    push(data.getBytes(charset))
  }
  def push(data: Array[Byte]): ZMessage = {
    _parts prepend data
    this
  }

  def pop() = {
    if (_parts.size > 0) new String(_parts.remove(0), charset)
    else ""
  }

  def addresses = {
    if (size > 5) _parts.dropRight(5).filterNot(_.isEmpty).toSeq
    else Seq.empty
  }

  def addresses_=(addr: Seq[Array[Byte]]) = {
    _parts = if (size > 5) { _parts.takeRight(5) } else _parts
    _parts prepend DELIMITER
    _parts.prepend(addr: _*)
    this
  }
  def append(data: Array[Byte]): ZMessage = {
    _parts.append(data)
    this
  }

  def unwrap() = {
    var address = _parts.remove(0)
    if (address.isEmpty) {
      address = _parts.remove(0)
    }
    if (_parts.headOption.forall(_.isEmpty)) {
      _parts.remove(0) // also remove the delimiter in one go
    }
    new String(address, charset)
  }

  def wrap(pts: String*) = {
    _parts.prepend(pts.map(_.getBytes(charset)): _*)
    this
  }

  def apply(socket: Socket) = {
    val mores = _parts.dropRight(1)
    val payload = _parts.takeRight(1).head
    mores foreach { socket.send(_, SendMore) }
    socket.send(payload, 0)
    this
  }

  override def toString = _parts.map(b ⇒ new String(b, charset)).mkString("ZMessage(", ", ", ")")
}

File: src/main/ruby/lib/backchat_borg/zmessage.rb
# -*- encoding: utf-8 -*-

module Backchat
  module Borg
    class ZMessage

      DELIMITER = ""

      def self.read(socket)
        parts = []
        socket.recv_strings parts
        ZMessage.new(*parts)
      end

      def initialize(*args)
        @parts = args
      end

      def size 
        @parts.size
      end

      def address=(addr)
        append addr
      end

      def address
        @parts.first || DELIMITER
      end

      def body
        @parts.last || DELIMITER
      end

      def body=(bd)
        @parts = @parts[0..-2]
        append bd
      end

      def send_to(socket)
        sender = socket.respond_to?(:raw_socket) ? socket.raw_socket : socket
        rc = 0
        @parts[0..-2].each { |part|
          rc = sender.send_string part, ZMQ::SNDMORE
        }
        rc = sender.send_string @parts[-1], 0
      end

      def to_s
        "ZMessage(#{@parts.map { |part| "'#{part}'" }.join(", ")})"
      end

      def append(data)
        @parts.push data
        self
      end

      def push(data)
        @parts.unshift data
        self
      end

      def pop
        @parts.shift || DELIMITER
      end

      def ccid
        get(5)
      end

      def ccid=(mt)
        set(5, mt)
      end

      def target 
        get(2)
      end

      def target=(mt)
        set(2, mt)
      end

      def sender
        get(3)
      end

      def sender=(mt)
        set(3, mt)
      end

      def message_type
        get(4)
      end

      def message_type=(mt) 
        set(4, mt)
      end

      def addresses
        (size > 5 ? @parts[0...-5].reject { |p| p.empty? } : []).flatten
      end

      def addresses=(*addr) 
        @parts = size > 5 ? @parts[0...5] : @parts
        @parts.unshift DELIMITER
        @parts = addr + @parts
        self
      end

      def unwrap
        addr = @parts.shift
        addr = @parts.shift if addr.empty?
        if (@parts.first||"").empty?
          @parts.shift # also remove the delimiter in one go
        end
        addr
      end

      def wrap(*pts) 
        @parts = pts + @parts
        self
      end

      def clone
        ZMessage.new(*@parts.clone)
      end

      private
      def get(index)
        @parts.size >= index ? @parts[size - index] || DELIMITER : DELIMITER
      end

      def set(index, value)
        if (size > (index - 1))
          @parts.delete_at(size - index)
          @parts.insert(size - (index - 1), value)
        else 
          push(value)
        end
      end
    end
  end
end
